clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
-                body.id().hir_id,
-            );
-        } else {
-            let mut rust_cc = cc + divergence - match_arms - short_circuits;
-            // prevent degenerate cases where unreachable code contains `return` statements
-            if rust_cc >= ret_adjust {
-                rust_cc -= ret_adjust;
-            }
-            if rust_cc > self.limit.limit() {
-                span_help_and_lint(
-                    cx,
-                    COGNITIVE_COMPLEXITY,
-                    span,
-                    &format!("the function has a cognitive complexity of {}", rust_cc),
-                    "you could split it up into multiple smaller functions",
-                );
-            }
-        }
-    }
-}
+                helper.visit_block(&fn_block);
+
+                let fn_score = helper.score as u64;
+                let score_limit = self.limit.limit();
 
-impl<'a, 'tcx> LateLintPass<'a, 'tcx> for CognitiveComplexity {
-    fn check_fn(
-        &mut self,
-        cx: &LateContext<'a, 'tcx>,
-        _: intravisit::FnKind<'tcx>,
-        _: &'tcx FnDecl,
-        body: &'tcx Body,
-        span: Span,
-        hir_id: HirId,
-    ) {
-        let def_id = cx.tcx.hir().local_def_id_from_hir_id(hir_id);
-        if !cx.tcx.has_attr(def_id, "test") {
-            self.check(cx, body, span);
+                if fn_score > score_limit {
+                    span_help_and_lint(
+                        cx,
+                        COGNITIVE_COMPLEXITY,
+                        item.span,
+                        &format!("the function has a cognitive complexity of {}", fn_score),
+                        "you could split it up into multiple smaller functions",
+                    );
+                }
+            }
         }
     }
 
-    fn enter_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn enter_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.push_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
-    fn exit_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn exit_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.pop_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
 }
 
-struct CCHelper<'a, 'tcx: 'a> {
-    match_arms: u64,
-    divergence: u64,
-    returns: u64,
-    short_circuits: u64, // && and ||
-    cx: &'a LateContext<'a, 'tcx>,
+struct CoCHelper {
+    current_nesting: u64,
+    score: f64,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for CCHelper<'a, 'tcx> {
-    fn visit_expr(&mut self, e: &'tcx Expr) {
-        match e.node {
-            ExprKind::Match(_, ref arms, _) => {
-                walk_expr(self, e);
-                let arms_n: u64 = arms.iter().map(|arm| arm.pats.len() as u64).sum();
-                if arms_n > 1 {
-                    self.match_arms += arms_n - 2;
-                }
+impl CoCHelper {
+    fn push_nesting(&mut self) {
```

Comment:
Fold these two methods into one method that takes a closure and does the increment/decrement logic around calling that closure.

Reply:
I'm sort of doing that in the `ExprKind::Block` branch. I like it because it doesn't need closures to work and is pretty simple. I fear that taking a closure as parameter would make this code much harder to "get" for a reader in the future who wishes to modify it. What do you think?

Reply:
(However, I might be underestimating how clear that code can get. I'll have a look at it nevertheless)

Reply:
> Are all nestings going to be of this kind?

I think they are. I'll think of it again, but atm of writing the implementation and checking the draft, it seemed to me that all nestings coincided with blocks (pairs of `{}`). That does not make the concept of nesting only possible within blocks, though. Refactoring the idea into a separate walking function that's independent of `ExprKind::Block` is definitely how it should look if we expand into it :D

Reply:
I can think of two main ways to nest something outside (figuratively) of a `Block`:

* One is when you have scoring expressions inside of a macro or otherwise a function call:

        `println!("The truth value is: {}", a && b || c);`

* The other is when you have parenthesis: `a + (...)`

The other ones would be these:

* After a `box`, however, this is unstable: `box Expr`

* Inside of a tuple: `(a, b, if c { 5 } else { - 6 })`

* While indexing something: `foo[...]`

* In a range: `a..(b + 5)`

* In an array definition: `[a, b, [c, d && e]]`